ksql-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectDataTranslator.java

```diff
@@ -52,60 +58,145 @@ public GenericRow toKsqlRow(final Schema connectSchema,
     );
   }
 
+  private RuntimeException createTypeMismatchException(final String pathStr,
+                                                       final Schema schema,
+                                                       final Schema connectSchema) {
+    throw new DataException(
+        String.format(
+            "Cannot deserialize type %s as type %s for field %s",
+            connectSchema.type().getName(),
+            schema.type().getName(),
+            pathStr));
+  }
+
+  private void validateType(final String pathStr,
+                            final Schema schema,
+                            final Schema connectSchema,
+                            final Schema.Type... validTypes) {
+    Arrays.asList(validTypes).stream()
+        .filter(connectSchema.type()::equals)
+        .findFirst()
+        .orElseThrow(
+            () -> createTypeMismatchException(pathStr, schema, connectSchema));
+  }
+
+  private void validateType(final String pathStr,
+                            final Schema schema,
+                            final Schema connectSchema) {
+    validateType(pathStr, schema, connectSchema, schema.type());
+  }
+
+  private void validateSchema(final String pathStr,
+                              final Schema schema,
+                              final Schema connectSchema) {
+    switch (schema.type()) {
+      case BOOLEAN:
+      case STRING:
+      case ARRAY:
+      case MAP:
+      case STRUCT:
+        validateType(pathStr, schema, connectSchema);
+        break;
+      case INT64:
+        validateType(
+            pathStr, schema, connectSchema,
+            Schema.Type.INT64, Schema.Type.INT32, Schema.Type.INT16, Schema.Type.INT8);
+        break;
+      case INT32:
+        validateType(
+            pathStr, schema, connectSchema,
+            Schema.Type.INT32, Schema.Type.INT16, Schema.Type.INT8);
+        break;
+      case FLOAT64:
+        validateType(pathStr, schema, connectSchema, Schema.Type.FLOAT32, Schema.Type.FLOAT64);
+        break;
+      default:
+        throw new RuntimeException(
```

Comment:
Should this be `UnsupportedTypeException` as in another one of your PRs?

Reply:
UnsupportedTypeException was intended to be internal to ConnectSchemaTranslator.